File: Algorithms/p215_Kth_Largest_Element_in_an_Array/p215_Kth_Largest_Element_in_an_Array.py
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        random.shuffle(nums)
        n = len(nums)
        l = 0
        r = n
        while True:
            x = nums[r-1]
            j = l-1
            for i in range(l,r-1):
                if nums[i]<x:
                    j += 1
                    nums[i],nums[j] = nums[j],nums[i]
            j += 1
            nums[j],nums[r-1] = nums[r-1],nums[j]
            if j<n-k:
                l = j+1
            elif j>n-k:
                r = j
            else:
                res = nums[j]
                break
        return res

File: Algorithms/p215_Kth_Largest_Element_in_an_Array/p215_Kth_Largest_Element_in_an_Array.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        random_shuffle(nums.begin(),nums.end());
        int n = nums.size();
        int l = 0;
        int r = n;
        int res;
        while(1){
            int x = nums[r-1];
            int j = l-1;
            for(int i=l;i<r;++i){
                if(nums[i]<x) swap(nums[i],nums[++j]);
            }
            swap(nums[++j],nums[r-1]);
            if(j<n-k){
                l = j+1;
            }else if(j>n-k){
                r = j;
            }else{
                res = nums[j];
                break;
            }
        }
        return res;
    }
};
